redis/asyncio/cluster.py

```diff
@@ -0,0 +1,1118 @@
+        elif host is not None and port is not None:
+            startup_nodes.append(ClusterNode(host, port))
+        elif len(startup_nodes) == 0:
+            # No startup node was provided
+            raise RedisClusterException(
+                "RedisCluster requires at least one node to discover the "
+                "cluster. Please provide one of the followings:\n"
+                "1. host and port, for example:\n"
+                " RedisCluster(host='localhost', port=6379)\n"
+                "2. list of startup nodes, for example:\n"
+                " RedisCluster(startup_nodes=[ClusterNode('localhost', 6379),"
+                " ClusterNode('localhost', 6378)])"
+            )
+        log.debug(f"startup_nodes : {startup_nodes}")
+        # Update the connection arguments
+        # Whenever a new connection is established, RedisCluster's on_connect
+        # method should be run
+        kwargs.update({"redis_connect_func": self.on_connect})
+        kwargs = cleanup_kwargs(**kwargs)
+
+        self.encoder = Encoder(
+            kwargs.get("encoding", "utf-8"),
+            kwargs.get("encoding_errors", "strict"),
+            kwargs.get("decode_responses", False),
+        )
+        self.cluster_error_retry_attempts = cluster_error_retry_attempts
+        self.command_flags = self.__class__.COMMAND_FLAGS.copy()
+        self.node_flags = self.__class__.NODE_FLAGS.copy()
+        self.read_from_replicas = read_from_replicas
+        self.reinitialize_counter = 0
+        self.reinitialize_steps = reinitialize_steps
+        self.nodes_manager = None
+        self.nodes_manager = NodesManager(
+            startup_nodes=startup_nodes,
+            from_url=from_url,
+            require_full_coverage=require_full_coverage,
+            **kwargs,
+        )
+
+        self.cluster_response_callbacks = CaseInsensitiveDict(
+            self.__class__.CLUSTER_COMMANDS_RESPONSE_CALLBACKS
+        )
+        self.result_callbacks = CaseInsensitiveDict(self.__class__.RESULT_CALLBACKS)
+        self.commands_parser = CommandsParser()
+        self._initialize = True
+        self._lock = asyncio.Lock()
+
+    async def initialize(self, force=False):
+        if self._initialize or force:
+            self._initialize = False
+            await self.nodes_manager.initialize()
+            await self.commands_parser.initialize(self)
+        return self
+
+    async def __aenter__(self):
+        return await self.initialize()
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        await self.close()
+
+    def __await__(self):
+        return self.initialize().__await__()
+
+    def __del__(self):
+        try:
+            loop = asyncio.get_event_loop()
+            coro = self.close()
+            if loop.is_running():
+                loop.create_task(coro)
+            else:
+                loop.run_until_complete(coro)
+        except Exception:
+            pass
```

Comment:
I'd rather have it like the client itself which is to get the user to manually close (take a look at the client del code). Thoughts?

Reply:
Agreed. It should also solve the CI issues. I'll update the PR.